### ISSUE TITLE
Reduce apphost template permissions in SDK tar.gz

### ISSUE THREAD

**Issue by dagood** [dotnet/core-setup#8863]
Right now, the apphost template file gives execute and write permissions to owner: `-rwxr--r-`. These aren't necessary, `-r--r--r-` would be better. See https://github.com/dotnet/core-setup/issues/7062#issuecomment-516936557 and convo in that thread. I believe this should also apply to the Deb/RPM packages.

**Comment by am11** [comment]
In my understanding; if we set this permission in toolset: https://github.com/dotnet/toolset/pull/3623, we do not need to make the change anywhere else. Is there any other reason to keep `u+x` permission in toolset?

**Comment by dagood** [comment]
After taking a look at https://github.com/dotnet/toolset/pull/3623, I think the missing link is the difference between template apphost and generated apphost (apphost shim). The template is put on disk by installers, the tar.gz, or NuGet packages. The apphost template is never executed so it doesn't need `u+x` (or anything else above `r`). It's read by the SDK during `dotnet publish` and some content is replaced to produce a generated apphost (apphost shim) that runs the app. The shim needs `x` so it can be executed.

**Comment by am11** [comment]
Is it possible to enforce the permission through .tar.gz? Seems like we can preserve file permission with `-p` in `.tar.gz`, but that is only the creation time fix. On extraction time, we still expect user to use `-p`. For example: https://github.com/dotnet/toolset/blob/3a43b10/scripts/obtain/dotnet-install.sh#L680.

**Comment by dagood** [comment]
> Is it possible to enforce the permission through .tar.gz? Seems like we can preserve file permission with `-p` in `.tar.gz`, but that is only the creation time fix.

I don't *think* we need to enforce it to that degree--if someone chooses to extract without preserving permissions, that's up to them, and they can also change permissions afterwards and I don't think we can go after that.

The concern I believe we have is someone running/installing .NET Core with a Microsoft-supported method and ending ending up with the template apphost on their disk owned by `root` with more permissions granted than are actually needed. /cc @vitek-karas @swaroop-sridhar 

> For example: https://github.com/dotnet/toolset/blob/3a43b10/scripts/obtain/dotnet-install.sh#L680.

Missing permission-preserving args in `dotnet-install.sh` seems borderline... we expect people to run that as a user, not root, so we may not have the same concern there. I'm not sure. (If we do fix, we have to be very careful it doesn't break some other scenario because of some assumption that no longer holds up.)

**Comment by omajid** [comment]
> These aren't necessary, `-r--r--r--` would be better.

I just realized that `-rx-rx-rx-` might be better, after all.

The native distro tooling that deals with debuginfo stripping (see https://github.com/dotnet/source-build/issues/267) only strips executibles. Making things `-r--r--r--` means that the tooling skips these binaries when it comes to handling debuginfo correctly.

**Comment by dagood** [comment]
Maybe a way to frame this is that the apphost template *isn't* executable, at least by the definition of "does something useful when executed"... it's only useful once the SDK makes the apphost shim using the template. It might make more sense to strip the symbols when the dev actually publishes their .NET Core app as a RPM/Deb package?

Something to consider is if the symbols on the apphost template "match" the bits on the generated apphost shim. If not, then stripping is the wrong thing to do either way, since the generated shims will not be debuggable at all. If that's not a problem, I suppose this discussion is about pragmatics:

* Template file size (in the apphost pack that everyone installs)
* Whether we're exposing security surface area by installing this template as `x` when it doesn't need it
* Whether we should go along with distro packaging tools' expectations (defaults?) to get an end result we want even though the conventions don't quite seem accurate vs. what these files truly are

I think we need more info from the host folks (@vitek-karas, @swaroop-sridhar, ...?) to know how to properly weigh this or get a firm requirement.

**Comment by vitek-karas** [comment]
We do need the host to be debuggable - not due to actual devs debugging it (although that also helps), but due to dump collection and analysis. If there are no symbols the automatic analysis gets thrown off and reports weird results.
Did we measure how much we save by stripping symbols?

I assume this only applies to Linux/Mac since on Windows the symbols are in a separate file.

**Comment by dagood** [comment]
To be concrete, my question is: can the *generated apphost shim* be debugged using symbols that were stripped from the *apphost template*? It sounds like yes, and we're already depending on that.

My suggestion above is exploring the idea that this assumption might not make sense, in particular in source-build. Maybe it's odd to think about it this way, but it seems like when the SDK replaces content in the apphost template, it's producing an executable file that "belongs" to the developer, so they might expect to own its symbols as well.

Maybe the replacements the SDK makes are small enough that we should think of this scenario as the dev redistributing a file that's *largely* unchanged.

---

@omajid is it possible to strip symbols manually when the file isn't detected by the packaging tooling? I'm sure it's not recommended--but at the root of all this, I believe we want to avoid `x` as a security measure.

> Did we measure how much we save by stripping symbols?

I wouldn't expect anyone to already have this measurement, since in the Microsoft build we strip symbols as a matter of course.

**Comment by vitek-karas** [comment]
I think the amount of modifications done by the SDK is small enough that the binary is still in spirit produces by Microsoft. It's kind of split:
* For purposes of ownership, servicing and security (signing and such) the modified binary produced by SDK is owned by the app developer.
* For purposes of diagnostics, crash dumps and so on - Microsoft authored all the code in it, so we own that. Not counting the fact that we currently don't have the tools necessary to produce a workable `.pdb` inside the SDK itself - so from a practical point of view, the symbols must be produced by the official build.

**Comment by dagood** [comment]
Ok, so I think we've arrived at a baseline: we don't believe the SDK can reasonably generate a debuggable apphost shim (whether that be generating a PDB, replacing content in a way that doesn't interfere with `.so` or `.dylib` symbols, etc.). The only place we think that can happen is when building the template.

That still leaves the dilemma from @omajid... install the apphost template as `-r--r--r--` for security and correctness like this issue tracks, or set it to `-rx-rx-rx-` so that source-build can (more easily?) produce a debuginfo package using normal RPM tooling?

---

> * For purposes of diagnostics, crash dumps and so on - Microsoft authored all the code in it, so we own that.

A little nitpicking here--there are non-Microsoft contributors, and Microsoft isn't the only org building, delivering, and potentially needing to debug .NET Core. Ideally we can choose something here that will also work for source-build.

**Comment by vitek-karas** [comment]
You're absolutely right - I simply wanted to say that whoever builds the "repo" sources is the "owner" of the debugging experience.

One more small correction - the SDK does today produce a debuggable apphost. The way it does that is that it takes the debuggable template produced by "repo build" and modifies it. The modifications are small enough that they don't "break" the debuggability.
